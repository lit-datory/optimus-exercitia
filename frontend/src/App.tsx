import "src/i18n/config"
import { RouterProvider } from "react-router-dom"
import { router } from "./routes"

function App() {
  return <RouterProvider router={router} />
}

export default App
