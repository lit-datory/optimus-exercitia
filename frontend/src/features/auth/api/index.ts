export * from "./auth"
export * from "./auth.mock"
