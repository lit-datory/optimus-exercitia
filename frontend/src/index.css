@import "tailwindcss";

@layer base {
  html {
    font-family: Poppins, system-ui, sans-serif;
  }
}
