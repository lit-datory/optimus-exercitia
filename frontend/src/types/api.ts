export type UrlParams = string | URLSearchParams | Record<string, string> | string[][]

export enum HttpMethods {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  PATCH = "PATCH",
  DELETE = "DELETE",
}
