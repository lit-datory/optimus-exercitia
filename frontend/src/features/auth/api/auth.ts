import { nonAuthorizedApi } from "src/lib/axios"
import { HttpMethods } from "src/types"
import { parseCookie } from "src/utils"
import { authResponseSchema } from "../schemas"
import { AuthResponse } from "../types"

export const login = async ({ email, password }: { email: string; password: string }) => {
  const response = await nonAuthorizedApi<AuthResponse>("/auth/login", {
    method: HttpMethods.POST,
    data: { email, password },
    withCredentials: true,
  })
  return authResponseSchema.parse(response.data)
}

export const logout = async () => {
  return await nonAuthorizedApi("/auth/logout", { method: HttpMethods.POST, withCredentials: true })
}

export const refreshToken = async () => {
  const csrfToken = parseCookie(document.cookie, "_csrf")
  const response = await nonAuthorizedApi<AuthResponse>("/auth/refresh_token", {
    headers: { "X-CSRF-TOKEN": csrfToken },
    method: HttpMethods.POST,
    withCredentials: true,
  })
  return authResponseSchema.parse(response.data)
}
