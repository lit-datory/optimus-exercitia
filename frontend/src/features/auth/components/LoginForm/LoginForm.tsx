import { useForm } from "react-hook-form"
import { useNavigate } from "react-router-dom"
import { zodResolver } from "@hookform/resolvers/zod"
import { Button } from "src/components/Button"
import { InputField } from "src/components/InputField"
import { loginFieldValuesSchema } from "../../schemas"
import { LoginFieldValues } from "../../types"
import { loginAtom } from "../../store"
import { useSetAtom } from "jotai"
import { useTranslation } from "react-i18next"

export const LoginForm = () => {
  const navigate = useNavigate()
  const { t } = useTranslation()
  const login = useSetAtom(loginAtom)

  const {
    register,
    handleSubmit,
    formState: { errors },
    setError,
  } = useForm<LoginFieldValues>({
    mode: "onBlur",
    resolver: zodResolver(loginFieldValuesSchema),
    defaultValues: {
      email: "",
      password: "",
    },
  })

  const handleLogin = async (data: LoginFieldValues) => {
    try {
      await login(data)
      return navigate("/")
    } catch (e: unknown) {
      setError("password", {
        message: t("login.inputFields.errors.password"),
      })
    }
  }

  const handleError = (message?: string) => {
    if (!message) return
    return t(message)
  }

  return (
    <form onSubmit={handleSubmit(handleLogin)} className="space-y-4">
      <InputField
        {...register("email")}
        label={t("common.inputFields.email.label")}
        name="email"
        type="email"
        error={handleError(errors.email?.message)}
        autoComplete="email"
      />
      <InputField
        {...register("password")}
        label={t("common.inputFields.password.label")}
        name="password"
        type="password"
        error={handleError(errors.password?.message)}
        autoComplete="current-password"
      />
      <Button data-testid="login-button" type="submit" variant="primary">
        {t("login.buttons.signIn")}
      </Button>
    </form>
  )
}
