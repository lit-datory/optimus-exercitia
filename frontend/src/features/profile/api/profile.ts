import { authorizedApi } from "src/lib/axios"
import { HttpMethods } from "src/types"
import { profileResponseSchema } from "../schemas"
import { ProfileResponse } from "../types"

export const getProfile = async () => {
  const response = await authorizedApi<ProfileResponse>("/profile", {
    method: HttpMethods.GET,
  })
  return profileResponseSchema.parse(response.data)
}
