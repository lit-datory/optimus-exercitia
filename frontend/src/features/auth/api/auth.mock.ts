import * as auth from "./auth"
import { AuthResponse } from "../types"
import { AxiosResponse } from "axios"

type Data = {
  resolve?: AuthResponse
  reject?: string
}

export const mockLogin = ({ resolve, reject }: Data) => {
  return vi.spyOn(auth, "login").mockReturnValue(
    new Promise((resolves, rejects) => {
      if (resolve) { resolves(resolve); return; }

      if (reject) { rejects(new Error(reject)); return; }
    }),
  )
}

export const mockLogout = () => {
  return vi.spyOn(auth, "logout").mockReturnValue(
    new Promise((resolves) => {
      resolves("successfully logged out!" as unknown as AxiosResponse);
    }),
  )
}

export const mockRefreshToken = ({ resolve, reject }: Data) => {
  return vi.spyOn(auth, "refreshToken").mockReturnValue(
    new Promise((resolves, rejects) => {
      if (resolve) { resolves(resolve); return; }

      if (reject) { rejects(new Error(reject)); return; }
    }),
  )
}
