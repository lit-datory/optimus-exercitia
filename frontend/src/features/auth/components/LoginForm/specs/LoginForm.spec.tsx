import { renderWithProviders, screen, fireEvent, waitFor, useNavigateMock } from "src/test"
import { mockLogin } from "../../../api"
import { LoginForm } from "../LoginForm"

describe("LoginForm component", () => {
  describe("Validations", () => {
    it("Correctly validates email", async () => {
      renderWithProviders(<LoginForm />)
      const emailInput = screen.getByTestId("email-input")
      const loginButton = screen.getByTestId("login-button")
      fireEvent.input(emailInput, { target: { value: "" } })
      fireEvent.click(loginButton)
      const errorMessage = await screen.findByTestId("email-error")
      expect(errorMessage).toMatchSnapshot()
    })

    it("Correctly validates password is required", async () => {
      renderWithProviders(<LoginForm />)
      const passwordInput = screen.getByTestId("password-input")
      const loginButton = screen.getByTestId("login-button")
      fireEvent.input(passwordInput, { target: { value: "" } })
      fireEvent.click(loginButton)
      const errorMessage = await screen.findByTestId("password-error")
      expect(errorMessage).toMatchSnapshot()
    })
  })

  describe("Sign in flow", () => {
    it("Successfully signs in", async () => {
      const mock = mockLogin({ resolve: { accessToken: "token" } })
      const mockNavigate = useNavigateMock()
      renderWithProviders(<LoginForm />)
      const emailInput = screen.getByTestId("email-input")
      const passwordInput = screen.getByTestId("password-input")
      fireEvent.input(emailInput, { target: { value: "test@example.com" } })
      fireEvent.input(passwordInput, { target: { value: "test" } })

      const loginButton = screen.getByTestId("login-button")
      fireEvent.click(loginButton)
      await waitFor(() => {
        expect(mock).toHaveBeenCalledWith({ email: "test@example.com", password: "test" })
      })
      await waitFor(() => {
        expect(mockNavigate).toHaveBeenCalledWith("/")
      })
    })

    it("Sets error when endpoint returns error", async () => {
      const mock = mockLogin({ reject: "Unauthorized" })
      renderWithProviders(<LoginForm />)
      const emailInput = screen.getByTestId("email-input")
      const passwordInput = screen.getByTestId("password-input")
      fireEvent.input(emailInput, { target: { value: "test@example.com" } })
      fireEvent.input(passwordInput, { target: { value: "test" } })

      const loginButton = screen.getByTestId("login-button")
      fireEvent.click(loginButton)
      await waitFor(() => {
        expect(mock).toHaveBeenCalledWith({ email: "test@example.com", password: "test" })
      })

      const errorMessage = await screen.findByTestId("password-error")
      expect(errorMessage).toMatchSnapshot()
    })
  })
})
