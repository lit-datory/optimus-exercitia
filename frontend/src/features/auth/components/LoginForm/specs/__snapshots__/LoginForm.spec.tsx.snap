// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm component > Sign in flow > Sets error when endpoint returns error 1`] = `
<span
  class="text-sm text-red-400"
  data-testid="password-error"
>
  Wrong email or password
</span>
`;

exports[`LoginForm component > Validations > Correctly validates email 1`] = `
<span
  class="text-sm text-red-400"
  data-testid="email-error"
>
  Invalid email
</span>
`;

exports[`LoginForm component > Validations > Correctly validates password is required 1`] = `
<span
  class="text-sm text-red-400"
  data-testid="password-error"
>
  Required
</span>
`;
